// pages/create-article/index.js
import React, {useContext, useEffect, useState} from 'react';
import {Redirect} from 'react-router-dom';

import ArticleForm from 'components/article-form';
import useFetch from 'hooks/use-fetch';
import {CurrentUserContext} from 'contexts/current-user';

const CreateArticle = () => {
    const apiUrl = '/articles';
    const [isSuccessfulSubmit, setIsSuccessfulSubmit] = useState(false);
    const [{response, error}, doFetch] = useFetch(apiUrl);
    const [currentUserState] = useContext(CurrentUserContext);
    console.log('createArticle', response);

    const onSubmit = article => {
        doFetch({
            method: 'post',
            data: {
                article
            }
        })
    };
    const initialValues = {
        title: '',
        description: '',
        body: '',
        tagList: ''
    };

    useEffect(() => {
        if (!response) {
            return
        }
        setIsSuccessfulSubmit(true);
    }, [response]);

    if (currentUserState.isLoggedIn === null) {
        return null
    }

    if (isSuccessfulSubmit || currentUserState.isLoggedIn === false) {
        return <Redirect to="/" />
    }

    return (
        <div>
            <ArticleForm
                onSubmit={onSubmit}
                initialValues={initialValues}
                errors={(error && error.errors) || {}}
            />
        </div>
    )
};

export default CreateArticle;
